zirgen/docs/05_Muxes.md

```diff
@@ -0,0 +1,187 @@
+# Muxes
+
+Besides the struct-like components discussed in the last section, the other
+primary mode of composing components is with muxes. Similar to how a struct is
+an `A` _and_ a `B` _and_ a `C`, a mux is _exactly one of_ an `A` _or_ a `B` _or_
+a `C`, much like sum types/unions/variants in other languages. We call each of
+these possibilities the _arms_ of the mux, and exactly one of a mux's arms is
+_active_ on a particular row of the execution trace. The active mux arm is
+determined by an array of field elements called the _selector_, which must be
+convertible to `Array<Val, N>` where `N` is the number of arms, and it must be
+that exactly one of the entries of the selector is nonzero.
```

Comment:
I think that exactly one selector has to be equal to `1`, and the rest have to be zero.

Reply:
I'm not sure what restrictions we impose in compilation, but if I understand the implementation correctly then mathematically it works to be anything nonzero, since it is multiplying each `constraint == 0` statement by the selector value on the left hand side and we're in a field (and so don't have zero divisors).

Reply:
My understanding was the same as Tim's... but are there cases where we pull a value out of a mux as a "dot product" of the selector and its values on all arms? I'm thinking something like this:
```
v := [s1, s2, s3] -> (v1, v2, v3);
-->
v := s1 * v1 + s2 * v2 + s3 * v3;
```

If the selector is one-hot, then this is a valid transformation, otherwise it isn't. Do we do this?

Reply:
When interpreting SwitchOp, we enforce that exactly one selector is `1` and the rest are `0`.   Lowering to Zll::IfOp, that restriction goes away.